src/main/java/org/opensearch/performanceanalyzer/rca/store/rca/hotshard/HotShardRca.java

```diff
@@ -106,108 +96,151 @@ private void consumeFlowUnit(
+                        metricType.toString(),
+                        e);
             }
         }
     }
 
-    private void consumeMetrics(
-            final Metric metric,
-            final HashMap<IndexShardKey, SlidingWindow<SlidingWindowData>> metricMap) {
+    private void consumeMetric(Metric metric, AllMetrics.OSMetrics metricType) {
         for (MetricFlowUnit metricFlowUnit : metric.getFlowUnits()) {
             if (metricFlowUnit.getData() != null) {
-                consumeFlowUnit(metricFlowUnit, metric.getClass().getName(), metricMap);
+                consumeFlowUnit(metricFlowUnit, metricType);
             }
         }
     }
 
-    private double fetchUsageValueFromMap(
-            HashMap<IndexShardKey, SlidingWindow<SlidingWindowData>> usageMap,
-            IndexShardKey indexShardKey) {
-        double value = 0;
-        if (usageMap.get(indexShardKey) != null) {
-            value = usageMap.get(indexShardKey).readAvg(TimeUnit.SECONDS);
+    private void checkAgainstThreshold(
+            MinMaxPriorityQueue<NamedSummarizedWindow> queue,
+            Map.Entry<IndexShardKey, SummarizedWindow> entry,
+            AllMetrics.OSMetrics metricType,
+            double threshold) {
+        // TODO : Check if null can break out here
+        double metricValue = entry.getValue().readAvgMetricValue(TimeUnit.SECONDS, metricType);
+        if (metricValue > threshold) {
+            queue.add(new NamedSummarizedWindow(entry.getValue(), entry.getKey()));
+
+            LOG.debug(
+                    "Hot Shard Identified, Shard : {} , metricValue = {} , metricThreshold = {}, metricType = {}",
+                    entry.getKey().toString(),
+                    metricValue,
+                    threshold,
+                    metricType.toString());
+        }
+    }
+
+    private void drainQueue(
+            MinMaxPriorityQueue<NamedSummarizedWindow> queue,
+            Map<IndexShardKey, SummarizedWindow> consumersToSend) {
+        while (!queue.isEmpty()) {
+            NamedSummarizedWindow candidate = queue.remove();
+            if (consumersToSend.containsKey(candidate.indexShardKey)) {
+                continue;
+            }
+            consumersToSend.put(candidate.indexShardKey, candidate.summarizedWindow);
         }
-        return value;
+    }
+
+    private HotNodeSummary summarize(
+            Map<IndexShardKey, SummarizedWindow> consumersToSend, InstanceDetails instanceDetails) {
+
+        HotNodeSummary nodeSummary =
+                new HotNodeSummary(
+                        instanceDetails.getInstanceId(), instanceDetails.getInstanceIp());
+
+        for (Map.Entry<IndexShardKey, SummarizedWindow> entry : consumersToSend.entrySet()) {
+
+            IndexShardKey indexShardKey = entry.getKey();
+            double avgCpuUtilization = entry.getValue().readAvgCpuUtilization(TimeUnit.SECONDS);
+            double avgHeapAllocRate = entry.getValue().readAvgHeapAllocRate(TimeUnit.SECONDS);
```

Comment:
Based on `drainQueue()` implementation, the `consumersToSend` will have either CPU or HeapAlloRate value, but never both.

I would recommend re-structuring the HotShardSummary as:

```
    private final String nodeId;
    private double resource_value;
    private double resource_threshold;
    private double resource_type;
```

Thoughts?

Reply:
Hey They will have both, all the time we deal with Summarized windows, even inside queues, so the information is perserved and we send all of it.
Let me know if you still want to shrink it down.